src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,135 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int ARRAY_MIN_CAPACITY = 10;
+    private Object[] array = new Object[ARRAY_MIN_CAPACITY];
+    private int size = 0;
+
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("");
+        }
+        if (index == array.length || size + 1 == array.length) {
+            array = grow();
+            System.arraycopy(array, index,
+                    array, index + 1,
+                    size - index);
+        } else if (index < size) {
+            System.arraycopy(array, index,
+                    array, index + 1,
+                    size - index);
+        }
+        array[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        Object[] arr = new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            arr[i] = list.get(i);
+            add((T) arr[i], size);
+        }
+    }
 
+    private Object[] grow() {
+        int oldCapacity = array.length;
+        int newCapacity = (int) (oldCapacity * 1.5);
+        return array = Arrays.copyOf(array, newCapacity);
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= 0 && index < size) {
+            return (T) array [index];
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " is not valid");
+        }
```

Comment:
```suggestion
        }
        throw new ArrayListIndexOutOfBoundsException("Index " + index + " is not valid");
```
якщо іф закінчується рітьорном елсе не потрібен